### PR TITLE
JournalDB updates: stop swallowing KeyError & keep diff if persist() errors

### PR DESCRIPTION
**Pull request by carver**
### What was wrong?

- `KeyError` was being swallowed during `persist()`, which caused some painful debugging
- On any kind of error during `persist()`, `JournalDB` was dropping all its changes on the floor (those that haven't yet been persisted, anyway)

(Pulled from #1735)

### How was it fixed?

- Track whether the wrapped DB has the key, and don't attempt delete if it was already missing (slight performance benefit, too)
- Catch exceptions and rebuild the pending diff of changes so that `persist()` can be reattempted after catching an exception

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

File: eth/db/journal.py [hunk 1]
```diff
@@ -211,11 +221,14 @@ def __setitem__(self, key: bytes, value: bytes) -> None:
...
         val = self.get(key)
-        return val is not None and val is not DELETED_ENTRY
+        return val is not None and val not in (ERASE_CREATED_ENTRY, DELETED_ENTRY)
 
     def __delitem__(self, key: bytes) -> None:
         self.latest[key] = DELETED_ENTRY
 
+    def local_delete(self, key: bytes) -> None:
+        self.latest[key] = ERASE_CREATED_ENTRY
+
```
Comment (lithp) [inline eth/db/journal.py]: nitpick: It's a little weird and unintuitive that `DELETED_ENTRY` and `ERASE_CREATED_ENTRY` have such different APIs. Something like dropping `__delitem__` and adding `delete(key: bytes, exists_in_db: bool = False)` might be easier for future readers to understand.

File: eth/db/journal.py [hunk 2]
```diff
@@ -20,8 +20,18 @@ class DeletedEntry:
...
 DELETED_ENTRY = DeletedEntry()
 
+# 1. key not in wrapped
+# 2. key created in journal
+# 3. key deleted
+ERASE_CREATED_ENTRY = DeletedEntry()
+
 
 class Journal(BaseDB):
     """
```
Comment (lithp) [inline eth/db/journal.py]: nitpick: 100% not part of this PR, but the docstring on `is_flattened` is wrong. If you're feeling generous now is as good a time as any to fix it :)

File: eth/db/journal.py [hunk 3]
```diff
@@ -290,9 +305,13 @@ def has_clear(self) -> bool:
         return self.journal.has_clear(self.journal.root_changeset_id)
 
     def __delitem__(self, key: bytes) -> None:
-        if key not in self.journal and key not in self.wrapped_db:
-            raise KeyError(key, "key could not be deleted in JournalDB, because it was missing")
-        del self.journal[key]
+        if key in self.wrapped_db:
```
Comment (lithp) [inline eth/db/journal.py]: Do you know if any callers skip the pipeline and talk to `wrapped_db` directly? `JournalDB` assumes not, and this could lead to some weird bugs. If a `ERASE` is pushed but the database later has the value then it won't be deleted, if a `DELETE` is pushed but the database later loses the value `persis()` will fail, etc.

File: eth/db/journal.py [hunk 4]
```diff
@@ -198,7 +208,7 @@ def __getitem__(self, key: bytes) -> Union[bytes, DeletedEntry]:    # type: igno
         # Ignored from mypy because of https://github.com/python/typeshed/issues/2078
         for changeset_id, changeset_data in reversed(self.journal_data.items()):
             if changeset_id in self._clears_at:
-                return DELETED_ENTRY
+                return ERASE_CREATED_ENTRY
```
[the remark this answers is not in the document]
Comment (lithp) [reply eth/db/journal.py]: Even more strangely:

```python
>>> memory_db = MemoryDB({b'one': b'one'})
>>> journal_db = JournalDB(memory_db)
>>> journal_db.clear()
>>> journal_db[b'one']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/brian/ef/py-evm/eth/db/journal.py", line 265, in __getitem__
    raise KeyError(key, "item was absent, then created and deleted, by JournalDB")
KeyError: (b'one', 'item was absent, then created and deleted, by JournalDB')
```

File: eth/db/journal.py [hunk 5]
```diff
@@ -376,14 +411,17 @@ def diff(self) -> DBDiff:
         visited_keys = set()  # type: Set[bytes]
 
         # Iterate in reverse, so you can skip over any keys from old checkpoints.
-        # This is purely for performance, not correctness.
+        # This is required so that when a key is created and then deleted in the journal,
+        #   we don't add the delete to the diff. (We simply omit the change altogether)
         for changeset in reversed(self.journal.journal_data.values()):
```
[the remark this answers is not in the document]
Comment (carver) [reply eth/db/journal.py]: Ah, thanks!

Interesting, I'm kind of shocked that this didn't trigger a consensus test failure, where storage elements are still available after deletion! They would have to be created just before deletion, but that seems like an obvious thing to check.

I don't understand the suggested solution (it seemed to error for me). But the test case was the important part. I've got it working now :+1:

---

> This is a weak code smell but eth/db/journal.py seems like it's waiting for an abstraction to be found. Starting with `self.journal.journal_data`, it's a little strange that `JournalDB` knows so much about how to read changesets when `Journal` already performs this operation internally.

Yeah, this whole module is a mess. The split of responsibilities doesn't make much sense to me.

Fair criticism that `diff()` is really dealing with `Journal` internals, for no good reason. I moved it inside `Journal`.

---

> One option might be: Instead of `Journal` maintaining an explicit list of changesets it holds onto the head (highest precidence) of a linked list of changesets. Each changeset handles `__getitem__` and `__delitem__` and such by looking first at itself and then recursing into the next changeset. `clear()` is easy to express, it's a subclass which doesn't recurse. Manipulating this stack might be easier that the current logic. For example, `_reapply_changeset_to_journal` could just pass the same stack back to Journal.

Yeah, this seems reasonable to me at first glance. But it's also a tangent of a tangent of a thing I'm trying to get done. As long as we can get it correct for now, I'm okay putting off a refactor for a bit.

File: eth/db/journal.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (carver) [reply eth/db/journal.py]: Ah, right, I'll try to make the error message a bit broader. I don't think it's worth tracking a third kind of delete (yet) :)

File: eth/db/journal.py (hunk 3 above)
Comment (carver) [reply eth/db/journal.py]: Yeah, I think this is broadly true of our database chains (for example, `CacheDB` will be confused if something it wraps is changed out-of-band). I think this is just something we live with in the current model of wrapping. It's the pipeliner's responsibility to not access the wrapped DB, or to have an understanding of how to fix it up if they do. (eg~ by calling `reset_cache()` on `CacheDB`)

File: eth/db/journal.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (carver) [reply eth/db/journal.py]: Oh yeah, your second one didn't come in on my feed before I answered. Hm, this makes it difficult to reliably say anything about a key when it's cleared. (The primary benefit of `clear()` is to not have to enumerate the underlying keys) I think I might have to just drop the custom message and raise a boring `KeyError`

**Comment by carver** [comment]
Ok, I'll go ahead and merge since I got the :heavy_check_mark: -- if there are any more follow-ups, I'll open a new one.

File: eth/db/journal.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lithp) [reply eth/db/journal.py]: > I don't understand the suggested solution (it seemed to error for me)

Ah, sorry, should have said the snippet wasn't a solution to the bug, but it might be a step closer to de-duplicating the changeset iteration logic.

> Yeah, this seems reasonable to me at first glance. But it's also a tangent of a tangent of a thing I'm trying to get done. As long as we can get it correct for now, I'm okay putting off a refactor for a bit.

And yeah, totally understood.

File: eth/db/journal.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lithp) [reply eth/db/journal.py]: The new message looks good! :+1:
